File: ago-dic-2024/Alfonso_Hernandez/Practica_4.java
package algoritmos;
import java.util.Scanner;

public class Practica_4 {
    
    // Método principal para ordenar la lista usando Merge Sort
    public static void mergeSort(String[] arr, int left, int right) {
        if (left < right) {
            int mid = (left + right) / 2;

            // Divide la lista en dos mitades
            mergeSort(arr, left, mid);
            mergeSort(arr, mid + 1, right);

            // Fusiona las dos mitades
            merge(arr, left, mid, right);
        }
    }

    // Método para fusionar dos mitades
    public static void merge(String[] arr, int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        // Arrays temporales para almacenar las dos mitades
        String[] leftArray = new String[n1];
        String[] rightArray = new String[n2];

        // Copia los datos a los arrays temporales
        for (int i = 0; i < n1; i++) {
            leftArray[i] = arr[left + i];
        }
        for (int i = 0; i < n2; i++) {
            rightArray[i] = arr[mid + 1 + i];
        }

        // Índices iniciales de las dos mitades y del array fusionado
        int i = 0, j = 0, k = left;

        // Ordena y fusiona los dos arrays temporales en el array original
        while (i < n1 && j < n2) {
            if (leftArray[i].compareTo(rightArray[j]) <= 0) {
                arr[k] = leftArray[i];
                i++;
            } else {
                arr[k] = rightArray[j];
                j++;
            }
            k++;
        }

        // Copia los elementos restantes de leftArray
        while (i < n1) {
            arr[k] = leftArray[i];
            i++;
            k++;
        }

        // Copia los elementos restantes de rightArray
        while (j < n2) {
            arr[k] = rightArray[j];
            j++;
            k++;
        }
    }

    public static void main(String[] args) {
        // Entrada de datos
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt(); // Número de palabras
        sc.nextLine(); // Consumir la línea sobrante
        String[] words = sc.nextLine().split(" "); // Lista de palabras

        // Llama al método mergeSort para ordenar la lista
        mergeSort(words, 0, words.length - 1);

        // Imprime las palabras ordenadas
        for (String word : words) {
            System.out.print(word + " ");
        }
        sc.close();
    }
}

File: ago-dic-2024/Zulema/Practica4.py
# Función de Merge Sort para ordenar las palabras
def merge_sort(words):
    if len(words) <= 1:
        return words
    
    # Dividir la lista en mitades
    mid = len(words) // 2
    left_half = merge_sort(words[:mid])
    right_half = merge_sort(words[mid:])
    
    # Mezclar las dos mitades ordenadas
    return merge(left_half, right_half)

# Función de mezcla para unir las dos mitades ordenadas
def merge(left, right):
    sorted_list = []
    i = j = 0
    
    # Comparar y mezclar
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            sorted_list.append(left[i])
            i += 1
        else:
            sorted_list.append(right[j])
            j += 1
    
    # Añadir los elementos restantes de cada mitad
    sorted_list.extend(left[i:])
    sorted_list.extend(right[j:])
    
    return sorted_list

# Lectura de entrada desde el estándar (STDIN)
def main():
    # Leer número de palabras (no se utiliza en el procesamiento)
    n = int(input().strip())
    
    # Leer la lista de palabras y dividirlas en una lista
    words = input().strip().split()
    
    # Ordenar la lista de palabras
    sorted_words = merge_sort(words)
    
    # Imprimir el resultado en formato requerido
    print(" ".join(sorted_words))

# Llamar a la función principal
if __name__ == "__main__":
    main()
